raiden-ts/src/path/epics.ts

```diff
@@ -1,48 +1,203 @@
   distinctUntilChanged,
+  filter,
+  first,
   groupBy,
-  switchMap,
+  map,
+  mergeMap,
+  pluck,
+  publishReplay,
   scan,
   startWith,
+  switchMap,
   tap,
-  delay,
+  timeout,
+  withLatestFrom,
 } from 'rxjs/operators';
 import { fromFetch } from 'rxjs/fetch';
-import { isActionOf, ActionType } from 'typesafe-actions';
-import { bigNumberify, BigNumber } from 'ethers/utils';
-import { Zero, Two } from 'ethers/constants';
+import { ActionType, isActionOf } from 'typesafe-actions';
+import { BigNumber, bigNumberify, hexDataLength, hexlify, toUtf8Bytes } from 'ethers/utils';
+import { Two, Zero } from 'ethers/constants';
 import { Event } from 'ethers/contract';
-import { isNil } from 'lodash';
+import { get, isNil } from 'lodash';
 
 import { RaidenAction } from '../actions';
 import { RaidenState } from '../state';
 import { RaidenEpicDeps } from '../types';
 import { getPresences$ } from '../transport/utils';
 import { messageGlobalSend } from '../messages/actions';
-import { PFSCapacityUpdate, MessageType } from '../messages/types';
-import { signMessage } from '../messages/utils';
+import { MessageType, PFSCapacityUpdate } from '../messages/types';
+import { MessageTypeId, signMessage } from '../messages/utils';
 import { channelDeposited } from '../channels/actions';
 import { ChannelState } from '../channels/state';
 import { channelAmounts } from '../channels/utils';
-import { Address, UInt, Int, decode } from '../utils/types';
-import { losslessStringify, losslessParse } from '../utils/data';
+import { Address, decode, Int, Signature, Signed, UInt } from '../utils/types';
+import { encode, losslessParse, losslessStringify } from '../utils/data';
 import { getEventsStream } from '../utils/ethers';
-import { pathFind, pathFound, pathFindFailed, pfsListUpdated } from './actions';
+import {
+  clearIOU,
+  pathFind,
+  pathFindFailed,
+  pathFound,
+  persistIOU,
+  pfsListUpdated,
+} from './actions';
 import { channelCanRoute, pfsInfo, pfsListInfo } from './utils';
-import { PathResults, Paths } from './types';
+import { IOU, LastIOUResults, PathResults, Paths, PFS } from './types';
+import { concat } from 'ethers/utils/bytes';
+import { Signer } from 'ethers';
+import { memoize } from 'lodash/fp';
+import { UserDeposit } from '../contracts/UserDeposit';
+import * as t from 'io-ts';
+
+/**
+ * Codec for PFS API returned error
+ */
+const PathError = t.readonly(
+  t.intersection([
+    t.type({
+      errors: t.string,
+      /* eslint-disable-next-line @typescript-eslint/camelcase */
+      error_code: t.number,
+    }),
+    t.partial({
+      /* eslint-disable-next-line @typescript-eslint/camelcase */
+      error_details: t.type({
+        from_: Address,
+        to: Address,
+        value: UInt(32),
+      }),
+    }),
+  ]),
+);
+interface PathError extends t.TypeOf<typeof PathError> {}
+
+type IOUPayload = {
+  serviceAddress: Address;
+  signedIOU?: Signed<IOU>;
+};
+
+class InvalidPFSRequestError extends Error {
```

Comment:
Nitpick: this would be the 2nd class in whole SDK. I've been avoiding this pattern, specially for errors, as errors should always be an exceptional case, not an expected one. Can this be refactored without the error class? Maybe an specific object or flag?

Reply:
This is at your discretion though, if you think it's worth it.

Reply:
Like we discussed, the main issue is that I have to perform a secondary action in some cases along with the failure. Addind a class extending error is the only way I can think of passing the iou to the error handling. But I am open to suggestions :+1: